Refactor (WebJS) - Fix some ESLint warnings

ShreyasJejurkar opened the pull request:
There are still some warnings that mostly say use `unknown` instead of `any`. 

Is there any particular reason why we use any instead of the unknown? 

And why use `unknown`? One of the potential reasons => https://mainawycliffe.dev/blog/typescript-use-unknown-instead-of-any/

https://github.com/typescript-eslint/typescript-eslint/blob/v4.32.0/packages/eslint-plugin/docs/rules/no-explicit-any.md

ShreyasJejurkar committed:
fix - Rename to `[FromService]` -> `[FromServices]`

fixes https://github.com/dotnet/aspnetcore/issues/38482

ShreyasJejurkar committed:
Revert "fix - Rename to `[FromService]` -> `[FromServices]`"

This reverts commit a08a629cbbb9c205d5f8229fd3fc47644b6151eb.

ShreyasJejurkar committed:
Merge branch 'dotnet:main' into main

ShreyasJejurkar committed:
Merge branch 'dotnet:main' into main

ShreyasJejurkar committed:
Merge branch 'dotnet:main' into main

ShreyasJejurkar committed:
Merge branch 'main' of https://github.com/dotnet/aspnetcore

ShreyasJejurkar committed:
Merge branch 'main' of https://github.com/ShreyasJejurkar/aspnetcore

ShreyasJejurkar committed:
Merge branch 'main' of https://github.com/ShreyasJejurkar/aspnetcore

ShreyasJejurkar committed:
Merge branch 'dotnet:main' into main

ShreyasJejurkar committed:
Merge branch 'dotnet:main' into main

ShreyasJejurkar committed:
Merge branch 'main' of https://github.com/dotnet/aspnetcore

ShreyasJejurkar committed:
Refactor (WebJS) - Fix some ESLint warnings

ShreyasJejurkar committed:
Merge branch 'dotnet:main' into main

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/LogicalElements.ts:
```diff
@@ -201,7 +201,8 @@ export function permuteLogicalChildren(parent: LogicalElement, permutationList:
   permutationList.forEach((listEntry: PermutationListEntryWithTrackingData) => {
-    const marker = listEntry.moveToBeforeMarker = document.createComment('marker');
+    const marker = document.createComment('marker');
+    listEntry.moveToBeforeMarker = document.createComment('marker');
```
This is a functional change as it's creating two separate comments. We should only create one.

ShreyasJejurkar replied:
Fixed with latest commit!

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/JSRootComponents.ts:
```diff
@@ -7,7 +7,7 @@ const pendingRootComponentContainerNamePrefix = '__bl-dynamic-root:';
 let nextPendingDynamicRootComponentIdentifier = 0;
 
-type ComponentParameters = object | null | undefined;
+type ComponentParameters = Record<string, unknown>
```
Is this definitely correct? What was wrong with it before? The previous definition was more specific about the potential for null/undefined which looks more relevant.

ShreyasJejurkar replied:
Yes. `object` is the ban type.
Reference -> https://github.com/typescript-eslint/typescript-eslint/blob/v4.32.0/packages/eslint-plugin/docs/rules/ban-types.md

ShreyasJejurkar replied:
`type ComponentParameters = {
    [x: string]: unknown;
}`
This is how the type has been inferred. The `unknown` can be `null` or `undefined`! I know it's not that verbose, if you think it might confuse people we can ignore this warning!

SteveSandersonMS replied:
I know the values inside the record can be null/undefined. The thing I'm talking about is that `ComponentParameters` *itself* can be null/undefined.

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/Renderer.ts:
```diff
@@ -17,7 +17,8 @@ let shouldResetScrollAfterNextBatch = false;
   if (!browserRenderer) {
-    browserRenderer = browserRenderers[browserRendererId] = new BrowserRenderer(browserRendererId);
+    browserRenderer = new BrowserRenderer(browserRendererId);
+    browserRenderers[browserRendererId] = new BrowserRenderer(browserRendererId);
```
This can't be right.

ShreyasJejurkar replied:
Fixed with the latest commit!

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/WebView/WebViewIpcSender.ts:
```diff
@@ -32,11 +32,11 @@ function base64EncodeByteArray(data: Uint8Array) {
     charBytes[i] = String.fromCharCode(data[i]);
   }
-  const dataBase64Encoded = btoa(charBytes.join(''));
+  const dataBase64Encoded = Buffer.from(charBytes.join(''), 'base64');
```
Are you sure this is correct? Can you confirm it's been tested in WebView?

ShreyasJejurkar replied:
I reverted this, seems like this is only available in NodeJS on the server side! Thanks for the catch!

SteveSandersonMS commented:
Thanks for submitting this. There are a few changes I think are needed, but once resolved, we'd be happy to take this PR!

TanayParikh commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -104,23 +104,23 @@ export const monoPlatform: Platform = {
   },
 
   readInt32Field: function readHeapInt32(baseAddress: Pointer, fieldOffset?: number): number {
-    return getValueI32((baseAddress as any as number) + (fieldOffset || 0));
```
This'll likely conflict with https://github.com/dotnet/aspnetcore/pull/38721.

ShreyasJejurkar commented:
I do have one question @SteveSandersonMS, is there any particular reason for using `any` instead of `unknown` in the current codebase.

ShreyasJejurkar committed:
fix : Addressed PR feedback.

ShreyasJejurkar committed:
Merge branch 'main' of https://github.com/dotnet/aspnetcore into js-improvements

ShreyasJejurkar committed:
fix : fix .eslintrc.js file

ShreyasJejurkar committed:
fix: webpack.config.js

SteveSandersonMS commented:
> is there any particular reason for using any instead of unknown in the current codebase

Mainly just that most of it was written before `unknown` was introduced in TypeScript v3

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Circuits/DefaultReconnectDisplay.ts:
```diff
@@ -72,11 +72,11 @@ export class DefaultReconnectDisplay implements ReconnectDisplay {
         // - false to mean we reached the server, but it rejected the connection (e.g., unknown circuit ID)
         // - exception to mean we didn't reach the server (this can be sync or async)
-        const successful = await Blazor.reconnect!();
+        const successful = await Blazor.reconnect?.();
```
This seems like a functional change. It's making the code tolerate `reconnect` being not set, whereas previously we were (in effect) asserting that it was set. Unless there's a scenario where being unset should be tolerated, could we revert this? Thanks!

ShreyasJejurkar replied:
Reverted!

SteveSandersonMS commented:
Thanks again @ShreyasJejurkar! There are just two remaining queries where I think the type annotations should be changed, but after that, I think we'll be ready to merge :)

ShreyasJejurkar committed:
Merge branch 'dotnet:main' into main

ShreyasJejurkar committed:
Merge branch 'main' of https://github.com/ShreyasJejurkar/aspnetcore into js-improvements

ShreyasJejurkar committed:
fix : Addressed PR feedback.

ShreyasJejurkar committed:
Reverted component parameters type

TanayParikh commented:
Thanks @ShreyasJejurkar for the changes. @SteveSandersonMS any concerns before merging?

SteveSandersonMS commented:
I was trying to review this yesterday but GitHub was giving me errors when adding comments on the code directly. Today it is still doing that, so I'm going to have to type in comments here:

* Loggers.ts line 11 - looks like we could just remove the empty constructor instead of annotating it. Not a big deal - I don't mind really.
* Platform.ts line 34-35 - the comment shouldn't have been changed; looks like a find-replace mistake (e.g., `without unknown boxing` should be reverted to `without any boxing`, and same for `unknown of these`). I do think this is worth fixing otherwise it becomes very strange.

So, very minor details, and sorry for not spotting these on the previous review.

ShreyasJejurkar commented:
@SteveSandersonMS Thanks for the review again! I appreciate! 

I am planning to open a separate PR to address these 2 issues, because I have deleted the fork of aspnetcore from my account, and along with that above branch is also deleted I guess, and I cannot pull this branch as it does not exist on my fork. 

@SteveSandersonMS @TanayParikh Will that be ok for both of you!? I will instantly open follow-up PR, once this get's merged!

SteveSandersonMS reviewed (approved):
Thanks!

SteveSandersonMS commented:
> @SteveSandersonMS @TanayParikh Will that be ok for both of you!? I will instantly open follow-up PR, once this get's merged!

Totally fine, thanks for checking. I'll merge now.

ShreyasJejurkar commented:
Ok, @SteveSandersonMS. Very thanks!

SteveSandersonMS commented:
Oh, bad news - because the remote branch has been deleted, it can't be merged or even checked out on my PC, as the branch no longer exists as far as GitHub is concerned. This might also explain the errors that GitHub was giving me yesterday when trying to post comments. Sorry!

I'm not sure if there's a way to get the code without manually recreating it. You might still have a local copy of the branch on your PC, which may be recoverable via `reflog` (e.g., https://stackoverflow.com/questions/3640764/can-i-recover-a-branch-after-its-deletion-in-git).

ShreyasJejurkar commented:
ohh gosh.....   

The problem is I am not using a PC for contributing. I use codespaces, and I think there isn't a way now to get back this one!

SteveSandersonMS commented:
It seems possible to SSH into a codespace machine, but I don't know it well enough to know whether you'd end up with the same state and would be able to recover a deleted Git branch that way: https://docs.github.com/en/codespaces/developing-in-codespaces/using-codespaces-with-github-cli#ssh-into-a-codespace.

If not, there might not be any alternative to manually recreating it.
